File: smash-ranks-android/app/src/main/java/com/garpr/android/activities/HeadToHeadActivity.java
package com.garpr.android.activities;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import com.garpr.android.App;
import com.garpr.android.R;
import com.garpr.android.adapters.HeadToHeadAdapter;
import com.garpr.android.misc.ListUtils;
import com.garpr.android.misc.RegionManager;
import com.garpr.android.misc.ThreadUtils;
import com.garpr.android.models.AbsPlayer;
import com.garpr.android.models.FullTournament;
import com.garpr.android.models.HeadToHead;
import com.garpr.android.models.Match;
import com.garpr.android.networking.ApiCall;
import com.garpr.android.networking.ApiListener;
import com.garpr.android.networking.ServerApi;
import com.garpr.android.views.ErrorLinearLayout;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;

public class HeadToHeadActivity extends BaseJavaActivity implements ApiListener<HeadToHead>,
        SwipeRefreshLayout.OnRefreshListener {

    private static final String TAG = "HeadToHeadActivity";
    private static final String CNAME = HeadToHeadActivity.class.getCanonicalName();
    private static final String EXTRA_PLAYER_ID = CNAME + ".PlayerId";
    private static final String EXTRA_PLAYER_NAME = CNAME + ".PlayerName";
    private static final String EXTRA_OPPONENT_ID = CNAME + ".OpponentId";
    private static final String EXTRA_OPPONENT_NAME = CNAME + ".OpponentName";

    private ArrayList<Object> mList;
    private HeadToHead mHeadToHead;
    private HeadToHeadAdapter mAdapter;
    private Match.Result mResult;
    private String mOpponentId;
    private String mOpponentName;
    private String mPlayerId;
    private String mPlayerName;

    @Inject
    RegionManager mRegionManager;

    @Inject
    ServerApi mServerApi;

    @Inject
    ThreadUtils mThreadUtils;

    @BindView(R.id.error)
    ErrorLinearLayout mError;

    @BindView(R.id.recyclerView)
    RecyclerView mRecyclerView;

    @BindView(R.id.refreshLayout)
    SwipeRefreshLayout mRefreshLayout;

    @BindView(R.id.empty)
    View mEmpty;


    public static Intent getLaunchIntent(final Context context, @NonNull final AbsPlayer player,
            @NonNull final AbsPlayer opponent) {
        return getLaunchIntent(context, player.getId(), player.getName(), opponent.getId(),
                opponent.getName());
    }

    public static Intent getLaunchIntent(final Context context, @NonNull final AbsPlayer player,
            @NonNull final Match match) {
        return getLaunchIntent(context, player.getId(), player.getName(),
                match.getOpponent().getId(), match.getOpponent().getName());
    }

    public static Intent getLaunchIntent(final Context context,
            @NonNull final FullTournament.Match match) {
        return getLaunchIntent(context, match.getWinnerId(), match.getWinnerName(),
                match.getLoserId(), match.getLoserName());
    }

    public static Intent getLaunchIntent(final Context context, @NonNull final String playerId,
            @Nullable final String playerName, @NonNull final String opponentId,
            @Nullable final String opponentName) {
        final Intent intent = new Intent(context, HeadToHeadActivity.class)
                .putExtra(EXTRA_PLAYER_ID, playerId)
                .putExtra(EXTRA_OPPONENT_ID, opponentId);

        if (!TextUtils.isEmpty(playerName) && !TextUtils.isEmpty(opponentName)) {
            intent.putExtra(EXTRA_PLAYER_NAME, playerName);
            intent.putExtra(EXTRA_OPPONENT_NAME, opponentName);
        }

        return intent;
    }

    @Override
    public void failure(final int errorCode) {
        mHeadToHead = null;
        mList = null;
        mResult = null;
        showError(errorCode);
    }

    private void fetchHeadToHead() {
        mRefreshLayout.setRefreshing(true);
        mServerApi.getHeadToHead(mRegionManager.getRegion(this), mPlayerId, mOpponentId,
                new ApiCall<>(this));
    }

    private void filter(@Nullable final Match.Result result) {
        mResult = result;

        if (mList == null || mList.isEmpty()) {
            return;
        }

        mThreadUtils.run(new ThreadUtils.Task() {
            private List<Object> mList;

            @Override
            public void onBackground() {
                if (!isAlive() || mResult != result) {
                    return;
                }

                mList = ListUtils.filterPlayerMatchesList(result, HeadToHeadActivity.this.mList);
            }

            @Override
            public void onUi() {
                if (!isAlive() || mResult != result) {
                    return;
                }

                mAdapter.set(mList);
                supportInvalidateOptionsMenu();
            }
        });
    }

    @Override
    protected String getActivityName() {
        return TAG;
    }

    @Override
    protected void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        App.get().getAppComponent().inject(this);
        setContentView(R.layout.activity_head_to_head);

        final Intent intent = getIntent();
        mOpponentId = intent.getStringExtra(EXTRA_OPPONENT_ID);
        mPlayerId = intent.getStringExtra(EXTRA_PLAYER_ID);

        if (intent.hasExtra(EXTRA_OPPONENT_NAME) && intent.hasExtra(EXTRA_PLAYER_NAME)) {
            mOpponentName = intent.getStringExtra(EXTRA_OPPONENT_NAME);
            mPlayerName = intent.getStringExtra(EXTRA_PLAYER_NAME);
            setSubtitle();
        }

        fetchHeadToHead();
    }

    @Override
    public boolean onCreateOptionsMenu(final Menu menu) {
        getMenuInflater().inflate(R.menu.activity_head_to_head, menu);

        if (mHeadToHead != null) {
            menu.findItem(R.id.miFilter).setVisible(true);
            menu.findItem(R.id.miFilterAll).setVisible(mResult != null);
            menu.findItem(R.id.miFilterLosses).setVisible(mResult != Match.Result.LOSE);
            menu.findItem(R.id.miFilterWins).setVisible(mResult != Match.Result.WIN);
        }

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        switch (item.getItemId()) {
            case R.id.miFilterAll:
                filter(null);
                return true;

            case R.id.miFilterLosses:
                filter(Match.Result.LOSE);
                return true;

            case R.id.miFilterWins:
                filter(Match.Result.WIN);
                return true;

            case R.id.miViewOpponent:
                startActivity(PlayerActivity.getLaunchIntent(this, mOpponentId, mOpponentName,
                        mRegionManager.getRegion(this)));
                return true;

            case R.id.miViewPlayer:
                startActivity(PlayerActivity.getLaunchIntent(this, mPlayerId, mPlayerName,
                        mRegionManager.getRegion(this)));
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onPrepareOptionsMenu(final Menu menu) {
        menu.findItem(R.id.miFilter).setVisible(mHeadToHead != null);

        if (!TextUtils.isEmpty(mOpponentName) && !TextUtils.isEmpty(mPlayerName)) {
            final MenuItem viewOpponent = menu.findItem(R.id.miViewOpponent);
            viewOpponent.setTitle(getString(R.string.view_x, mOpponentName));
            viewOpponent.setVisible(true);

            final MenuItem viewPlayer = menu.findItem(R.id.miViewPlayer);
            viewPlayer.setTitle(getString(R.string.view_x, mPlayerName));
            viewPlayer.setVisible(true);
        }

        return super.onPrepareOptionsMenu(menu);
    }

    @Override
    public void onRefresh() {
        fetchHeadToHead();
    }

    @Override
    protected void onViewsBound() {
        super.onViewsBound();

        mRefreshLayout.setOnRefreshListener(this);
        mRecyclerView.setHasFixedSize(true);
        mAdapter = new HeadToHeadAdapter(this);
        mRecyclerView.setAdapter(mAdapter);
    }

    private void prepareMenuAndSubtitle() {
        if (mHeadToHead != null) {
            if (TextUtils.isEmpty(mOpponentName)) {
                mOpponentName = mHeadToHead.getOpponent().getName();
            }

            if (TextUtils.isEmpty(mPlayerName)) {
                mPlayerName = mHeadToHead.getPlayer().getName();
            }
        }

        setSubtitle();
        supportInvalidateOptionsMenu();
    }

    private void showData() {
        mList = ListUtils.createHeadToHeadList(this, mHeadToHead);
        mAdapter.set(mList);
        mEmpty.setVisibility(View.GONE);
        mError.setVisibility(View.GONE);
        mRecyclerView.setVisibility(View.VISIBLE);
        prepareMenuAndSubtitle();
        mRefreshLayout.setRefreshing(false);
    }

    private void showEmpty() {
        mAdapter.clear();
        mRecyclerView.setVisibility(View.GONE);
        mError.setVisibility(View.GONE);
        mEmpty.setVisibility(View.VISIBLE);
        prepareMenuAndSubtitle();
        mRefreshLayout.setRefreshing(false);
    }

    private void showError(final int errorCode) {
        mAdapter.clear();
        mRecyclerView.setVisibility(View.GONE);
        mEmpty.setVisibility(View.GONE);
        mError.setVisibility(View.VISIBLE, errorCode);
        prepareMenuAndSubtitle();
        mRefreshLayout.setRefreshing(false);
    }

    @Override
    protected boolean showUpNavigation() {
        return true;
    }

    private void setSubtitle() {
        if (TextUtils.isEmpty(getSubtitle()) && !TextUtils.isEmpty(mPlayerName) &&
                !TextUtils.isEmpty(mOpponentName)) {
            setSubtitle(getString(R.string.x_vs_y, mPlayerName, mOpponentName));
        }
    }

    @Override
    public void success(@Nullable final HeadToHead headToHead) {
        mHeadToHead = headToHead;

        if (mHeadToHead == null) {
            showEmpty();
        } else {
            showData();
        }
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/activities/HeadToHeadActivity.kt
package com.garpr.android.activities

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.garpr.android.R
import com.garpr.android.adapters.HeadToHeadAdapter
import com.garpr.android.extensions.appComponent
import com.garpr.android.extensions.putOptionalExtra
import com.garpr.android.extensions.requireStringExtra
import com.garpr.android.managers.RegionManager
import com.garpr.android.misc.ListUtils
import com.garpr.android.misc.ThreadUtils
import com.garpr.android.models.AbsPlayer
import com.garpr.android.models.FavoritePlayer
import com.garpr.android.models.FullTournament
import com.garpr.android.models.HeadToHead
import com.garpr.android.models.Match
import com.garpr.android.models.Region
import com.garpr.android.networking.ApiCall
import com.garpr.android.networking.ApiListener
import com.garpr.android.networking.ServerApi
import kotlinx.android.synthetic.main.activity_head_to_head.*
import javax.inject.Inject

class HeadToHeadActivity : BaseActivity(), ApiListener<HeadToHead>,
        SwipeRefreshLayout.OnRefreshListener {

    private var list: List<Any>? = null
    private var headToHead: HeadToHead? = null
    private lateinit var adapter: HeadToHeadAdapter

    private val opponentId: String by lazy { intent.requireStringExtra(EXTRA_OPPONENT_ID) }

    @Inject
    protected lateinit var regionManager: RegionManager

    @Inject
    protected lateinit var serverApi: ServerApi

    @Inject
    protected lateinit var threadUtils: ThreadUtils


    companion object {
        private const val TAG = "HeadToHeadActivity"
        private val CNAME = HeadToHeadActivity::class.java.canonicalName
        private val EXTRA_PLAYER_ID = "$CNAME.PlayerId"
        private val EXTRA_OPPONENT_ID = "$CNAME.OpponentId"

        fun getLaunchIntent(context: Context, player: AbsPlayer, opponent: AbsPlayer,
                region: Region? = null): Intent {
            var regionCopy = region

            if (player is FavoritePlayer) {
                regionCopy = player.region
            }

            return getLaunchIntent(context, player.id, opponent.id, regionCopy)
        }

        fun getLaunchIntent(context: Context, player: AbsPlayer, match: Match,
                region: Region? = null): Intent {
            var regionCopy = region

            if (player is FavoritePlayer) {
                regionCopy = player.region
            }

            return getLaunchIntent(context, player.id, match.opponent.id, regionCopy)
        }

        fun getLaunchIntent(context: Context, match: FullTournament.Match,
                region: Region? = null): Intent {
            return getLaunchIntent(context, match.winnerId, match.loserId, region)
        }

        fun getLaunchIntent(context: Context, playerId: String, opponentId: String,
                region: Region? = null): Intent {
            return Intent(context, HeadToHeadActivity::class.java)
                    .putExtra(EXTRA_PLAYER_ID, playerId)
                    .putExtra(EXTRA_OPPONENT_ID, opponentId)
                    .putOptionalExtra(EXTRA_REGION, region)
        }
    }

    override val activityName = TAG

    override fun failure(errorCode: Int) {
        headToHead = null
        list = null
        showError(errorCode)
    }

    private fun fetchHeadToHead() {
        refreshLayout.isRefreshing = true
        serverApi.getHeadToHead(regionManager.getRegion(this), playerId, opponentId,
                ApiCall(this))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        appComponent.inject(this)
        setContentView(R.layout.activity_head_to_head)
        toolbar.subtitleText = regionManager.getRegion(this).displayName

        fetchHeadToHead()
    }

    override fun onRefresh() {
        fetchHeadToHead()
    }

    override fun onViewsBound() {
        super.onViewsBound()

        refreshLayout.setOnRefreshListener(this)
        recyclerView.setHasFixedSize(true)
        adapter = HeadToHeadAdapter(this)
        recyclerView.adapter = adapter
    }

    private val playerId: String by lazy { intent.requireStringExtra(EXTRA_PLAYER_ID) }

    private fun showData() {
        list = ListUtils.createHeadToHeadList(this, headToHead)
        adapter.set(list)

        if (adapter.isEmpty) {
            error.visibility = View.GONE
            recyclerView.visibility = View.GONE
            empty.visibility = View.VISIBLE
        } else {
            empty.visibility = View.GONE
            error.visibility = View.GONE
            recyclerView.visibility = View.VISIBLE
        }

        refreshLayout.isRefreshing = false
        toolbar.refresh()
    }

    private fun showError(errorCode: Int) {
        adapter.clear()
        empty.visibility = View.GONE
        recyclerView.visibility = View.GONE
        error.setVisibility(View.VISIBLE, errorCode)
        refreshLayout.isRefreshing = false
        toolbar.refresh()
    }

    override fun success(`object`: HeadToHead?) {
        headToHead = `object`
        list = null
        showData()
    }

}
